# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: js/src/jit/BaselineCacheIRCompiler.cpp
# Commit: 655698c18737
# Full Hash: 655698c18737242eda9190ce601a19b54c3826ef
# Author: Iain Ireland <iireland@mozilla.com>
# Date: 2020-09-18 09:45:00
# Description:
#   Bug 1665432: Avoid clobbering code register when calling CreateThis r=jandem
#   
#   Relative to `CallScriptedFunction`, `CallInlinedFunction` has an extra register (`codeReg`) alive when calling `CreateThis`, which was not being saved. On x64 we got lucky and it was assigned to a preserved register, but x86 broke.
#   
#   I took the opportunity to do a little bit more cleanup of `CreateThis` and document the justification for reloading `calleeReg`.
# ==============================================================================

diff -r 5b344e00670b -r 655698c18737 js/src/jit/BaselineCacheIRCompiler.cpp
--- a/js/src/jit/BaselineCacheIRCompiler.cpp	Thu Sep 17 16:10:12 2020 +0000
+++ b/js/src/jit/BaselineCacheIRCompiler.cpp	Thu Sep 17 06:50:04 2020 +0000
@@ -3183,7 +3183,8 @@
  * overwrites the magic ThisV on the stack.
  */
 void BaselineCacheIRCompiler::createThis(Register argcReg, Register calleeReg,
-                                         Register scratch, CallFlags flags) {
+                                         Register scratch, CallFlags flags,
+                                         LiveGeneralRegisterSet liveNonGCRegs) {
   MOZ_ASSERT(flags.isConstructing());
 
   if (flags.needsUninitializedThis()) {
@@ -3193,9 +3194,9 @@
 
   size_t depth = STUB_FRAME_SIZE;
 
-  // Save argc before call.
-  masm.push(argcReg);
-  depth += sizeof(size_t);
+  // Save live registers that don't have to be traced.
+  masm.PushRegsInMask(liveNonGCRegs);
+  depth += sizeof(uintptr_t) * liveNonGCRegs.set().size();
 
   // CreateThis takes two arguments: callee, and newTarget.
 
@@ -3222,17 +3223,16 @@
   masm.bind(&createdThisOK);
 #endif
 
-  // Restore argc
-  masm.pop(argcReg);
+  // Restore saved registers.
+  masm.PopRegsInMask(liveNonGCRegs);
 
   // Save |this| value back into pushed arguments on stack.
+  MOZ_ASSERT(!liveNonGCRegs.aliases(JSReturnOperand));
   storeThis(JSReturnOperand, argcReg, flags);
 
-  // Restore the stub register from the baseline stub frame.
-  Address stubRegAddress(masm.getStackPointer(), STUB_FRAME_SAVED_STUB_OFFSET);
-  masm.loadPtr(stubRegAddress, ICStubReg);
-
-  // Restore calleeReg.
+  // Restore calleeReg. CreateThisFromIC may trigger a GC, so we reload the
+  // callee from the stub frame (which is traced) instead of spilling it to
+  // the stack.
   depth = STUB_FRAME_SIZE;
   loadStackObject(ArgumentKind::Callee, flags, depth, argcReg, calleeReg);
 }
@@ -3294,7 +3294,10 @@
   }
 
   if (isConstructing) {
-    createThis(argcReg, calleeReg, scratch, flags);
+    LiveGeneralRegisterSet liveNonGCRegs;
+    liveNonGCRegs.add(argcReg);
+    liveNonGCRegs.add(ICStubReg);
+    createThis(argcReg, calleeReg, scratch, flags, liveNonGCRegs);
   }
 
   pushArguments(argcReg, calleeReg, scratch, scratch2, flags,
@@ -3349,8 +3352,8 @@
   JitSpew(JitSpew_Codegen, "%s", __FUNCTION__);
   AutoOutputRegister output(*this);
   AutoScratchRegisterMaybeOutput scratch(allocator, masm, output);
-  AutoScratchRegisterMaybeOutputType codeReg(allocator, masm, output);
-  AutoScratchRegister scratch2(allocator, masm);
+  AutoScratchRegisterMaybeOutputType scratch2(allocator, masm, output);
+  AutoScratchRegister codeReg(allocator, masm);
 
   Register calleeReg = allocator.useRegister(masm, calleeId);
   Register argcReg = allocator.useRegister(masm, argcId);
@@ -3381,7 +3384,11 @@
   }
 
   if (isConstructing) {
-    createThis(argcReg, calleeReg, scratch, flags);
+    LiveGeneralRegisterSet liveNonGCRegs;
+    liveNonGCRegs.add(argcReg);
+    liveNonGCRegs.add(ICStubReg);
+    liveNonGCRegs.add(codeReg);
+    createThis(argcReg, calleeReg, scratch, flags, liveNonGCRegs);
   }
 
   pushArguments(argcReg, calleeReg, scratch, scratch2, flags,
@@ -3429,7 +3436,7 @@
   stubFrame.leave(masm, true);
 
   if (!isSameRealm) {
-    masm.switchToBaselineFrameRealm(scratch2);
+    masm.switchToBaselineFrameRealm(codeReg);
   }
 
   return true;